include/clang/AST/ASTStructuralEquivalence.h

```diff
@@ -33,14 +33,12 @@ struct StructuralEquivalenceContext {
   /// AST contexts for which we are checking structural equivalence.
   ASTContext &FromCtx, &ToCtx;
 
-  /// The set of "tentative" equivalences between two canonical
-  /// declarations, mapping from a declaration in the first context to the
-  /// declaration in the second context that we believe to be equivalent.
-  llvm::DenseMap<Decl *, Decl *> TentativeEquivalences;
+  /// Queue of declaration (from, to) pairs whose equivalence still needs to
+  /// be verified.
+  std::deque<std::pair<Decl *, Decl *>> DeclsToCheck;
 
-  /// Queue of declarations in the first context whose equivalence
-  /// with a declaration in the second context still needs to be verified.
-  std::deque<Decl *> DeclsToCheck;
+  /// Set of those pairs which we already had checked for equivalency.
+  llvm::DenseSet<std::pair<Decl *, Decl *>> Visited;
```

Comment:
Are these sets `DeclsToCheck`, `Visited`, `NonEquivalentDecls` not storing a state of a single (Decl*, Decl*) pair?
If yes it can be better to have a map from a (Decl*, Decl*) pair to a state that can be: not-checked-yet (but visited), checked-but-not-equivalent, maybe-equivalent (visited and not in `NonEquivalentDecls`)?